include/RAJA/exec-cuda/forall_cuda.hxx

```diff
@@ -175,18 +206,26 @@ template <size_t BLOCK_SIZE, typename LOOP_BODY>
 RAJA_INLINE void forall(cuda_exec<BLOCK_SIZE>,
                         Index_type begin,
                         Index_type end,
-                        LOOP_BODY loop_body)
+                        LOOP_BODY&& loop_body)
 {
+  beforeCudaKernelLaunch();
+
+  auto body = loop_body;
+
   Index_type len = end - begin;
-  size_t gridSize = (len + BLOCK_SIZE - 1) / BLOCK_SIZE;
+  size_t gridSize =
+      RAJA_MIN((len + BLOCK_SIZE - 1) / BLOCK_SIZE, RAJA_CUDA_MAX_NUM_BLOCKS);
```

Comment:
This calculation of gridSize assumes a loop in the kernel to get all work done, which is not included. This should be changed to an assert on gridSize or the loop over len should be added back into the kernels.